### PR TITLE
Implement ROA export

### LINKED ISSUE #219: Extended rtrlclient to create multiple outputs
#### Background
Multiple implementations of RPKI origin validation (e.g., OpenBGPD) don't use RPKI RTR to maintain valid ROA data but import data from external files. Therefore, some RPKI validators provide a corresponding output which is then fed to the BGP router on a regular basis.

Such validator-centric approach is not good operational practice as it binds ROA output to a single Validator implementation and running instance. It would be better to benefit from the RPKI architecture, which allows a BGP router to receive data from multiple relying parties, e.g., to improve reliability.

#### Proposal (1)
The proposal is to extend the rtrclient such that it creates multiple output formats (see below) of the received valid ROA payload. It seems also helpful to introduce an option that terminates the rtrclient after receiving all data to make it easier to use in cron settings.

#### Proposal (2)
Preferably, we allow for easy extension of the output format, i.e., the user can configure how the output should look like. Typically, the required output consists of a fix preamble, dynamic content that includes the ROA, and a postamble.

For example, the OpenBGPD config:
```
roa-set { 
    203.119.88.0/23 maxlen 24 source-as 187
    2001:dd8:7::/48 maxlen 48 source-as 187        
    2401:680::/32 maxlen 32 source-as 715
}
deny quick from ebgp ovs invalid
```

A configuration file for this might look like
```
peamble:
  roa-set { 
content:
  $prefix/$length maxlen $maxlen source-as $origin
postamble:
  }
deny quick from ebgp ovs invalid
```

For a CSV file
```
  preamble:
    #this is a ROA CSV file
  content:
   $prefix,$length,$maxlen,$origin
  postamble:
```

### Useful links
https://rpki.readthedocs.io/en/latest/routinator/running.html#printing-a-list-of-valid-route-origins

### PR DESCRIPTION
**Pull request by mroethke**
### Contribution description

Implement export of ROA information through rtrclient in arbitrary formats.

### Issues/PRs references

Fix #219

### CODE REVIEW

**Comment by smlng** [comment]
without diving into this too deep, but from first thought: can't we integrate the `mustach` library differently than copying the whole project (including files like `.gitignore` and `.gitlab-ci.yml`). For instance by using CMAKE external projects, see [here](https://cmake.org/cmake/help/v3.8/module/ExternalProject.html)?

**Comment by waehlisch** [comment]
Don't make it complicate in terms of tooling.

**Comment by mroethke** [comment]
I removed of the mustachs files, keeping only what we use + license, readme and authors files.

The remaining travis issue is fixed by #242, which should probably be merged first.

File: tools/rtrclient.c [hunk 1]
```diff
@@ -98,6 +116,60 @@ static struct socket_config *extend_socket_config()
...
+
+	if (is_readable_file(name)) {
+		FILE *template_file = fopen(name, "r");
+
+		if (fseek(template_file, 0, SEEK_END) == -1)
+			goto read_error;
+
+		long file_size = ftell(template_file);
+		if (file_size == -1)
+			goto read_error;
```
Comment (smlng) [inline tools/rtrclient.c]: close file?

File: tools/rtrclient.c [hunk 2]
```diff
@@ -98,6 +116,60 @@ static struct socket_config *extend_socket_config()
...
+
+		rewind(template_file);
+
+		char *template = checked_malloc(file_size);
+
+		size_t bytes_read = fread(template, 1, file_size, template_file);
+
+		if (bytes_read != (unsigned long) file_size)
+			goto read_error;
```
Comment (smlng) [inline tools/rtrclient.c]: close file?

File: tools/rtrclient.c [hunk 3]
```diff
@@ -140,19 +212,144 @@ static bool is_resolveable_host(const char *str)
 	return true;
 }
 
-#ifdef RTRLIB_HAVE_LIBSSH
+static bool is_file(const char *str)
+{
+	struct stat stat_buf;
+
+	if (stat(str, &stat_buf) == -1 )
```
Comment (smlng) [inline tools/rtrclient.c]: remove space after `-1`

Comment (smlng) [reply tools/rtrclient.c]: travis (check patch in there) found that too and one or two more

File: tools/rtrclient.c [hunk 4]
```diff
@@ -169,6 +367,10 @@ static void print_usage(char **argv)
 	printf("-k  Print information about SPKI updates.\n");
 	printf("-p  Print information about PFX updates.\n\n");
 
+	printf("-e export pfx table and exit\n");
```
Comment (smlng) [inline tools/rtrclient.c]: these are not reflected/shown in `usage` nor `examples` in the help message.

Also: please change port from `8282` to `8283` in examples to utilise the RIPE Validator v3

File: tools/rtrclient.c (hunk 1 above)
Comment (mroethke) [reply tools/rtrclient.c]: done

File: tools/rtrclient.c (hunk 2 above)
Comment (mroethke) [reply tools/rtrclient.c]: same

File: tools/rtrclient.c (hunk 3 above)
Comment (mroethke) [reply tools/rtrclient.c]: already done

File: tools/rtrclient.c (hunk 4 above)
Comment (mroethke) [reply tools/rtrclient.c]: I added examples for the export functionality and documented the `-o` option.

File: third-party/CMakeLists.txt [hunk 5]
```diff
@@ -1 +1,4 @@
 set(tommyds "${CMAKE_CURRENT_SOURCE_DIR}/tommyds/tommy.c" PARENT_SCOPE)
+
+set(mustach "${CMAKE_CURRENT_SOURCE_DIR}/mustach/mustach.c" PARENT_SCOPE)
+set_source_files_properties("${CMAKE_CURRENT_SOURCE_DIR}/mustach/mustach.c" PROPERTIES COMPILE_FLAGS "-DNO_EXTENSION_FOR_MUSTACH")
```
Comment (smlng) [inline third-party/CMakeLists.txt]: why is this set here and again in `tools/CMakeLists.txt`, are both required? Btw. why not for tommyds then?

File: tools/rtrclient.c [hunk 6]
```diff
@@ -167,15 +368,21 @@ static void print_usage(char **argv)
 	printf("-b  bindaddr Hostnamne or IP address to connect from\n\n");
 
 	printf("-k  Print information about SPKI updates.\n");
-	printf("-p  Print information about PFX updates.\n\n");
+	printf("-p  Print information about PFX updates.\n");
+	printf("-s  Print information about connection status updates.\n\n");
```
Comment (smlng) [inline tools/rtrclient.c]: this is missing in usage above

File: third-party/CMakeLists.txt (hunk 5 above)
Comment (mroethke) [reply third-party/CMakeLists.txt]: I probably just to forgot to remove that one …

**Comment by mroethke** [comment]
not because no format is defined, as there is a default format, but because `-e` is not defined. Same for `-t`.

**Comment by mroethke** [comment]
That is exactly what it did when I entered the command just now

**Comment by smlng** [comment]
now it prints usage instead of running, however it does not state why its failing - can we add a reasonable error message (`invalid template`) before printing the help.

```
rtrclient -e -t foo -o roa.foo tcp rpki-validator.realmv6.org 8283
Usage:
 rtrclient tcp [options] <host> <port>
 rtrclient ssh [options] <host> <port> <username> <private_key> [<host_key>]

Options:
-k  Print information about SPKI updates.
-p  Print information about PFX updates.

Examples:
 rtrclient tcp rpki-validator.realmv6.org 8282
 rtrclient tcp -k -p rpki-validator.realmv6.org 8282
 rtrclient ssh rpki-validator.realmv6.org 22 rtr-ssh ~/.ssh/id_rsa ~/.ssh/known_hosts
 rtrclient ssh -k -p rpki-validator.realmv6.org 22 rtr-ssh ~/.ssh/id_rsa ~/.ssh/known_hosts
```

**Comment by smlng** [comment]
please squash fixups

**Comment by smlng** [comment]
tested ACK, please squash

**Comment by mroethke** [comment]
done

**Pull request merged**